File: TypeScript/bezier_curve.ts
function comb(n: number, k: number): number {
    // Calculate the binomial coefficient (n choose k)
    if (k > n) return 0;
    let res = 1;
    for (let i = 0; i < k; i++) {
        res *= (n - i) / (i + 1);
    }
    return res;
}

function bezierCurve(points: number[][], numFrames: number): [number[], number[]] {
    const n = points.length - 1;
    const t_vals = Array.from({ length: numFrames }, (_, i) => Math.pow(i / (numFrames - 1), 1.5));
    const curveX: number[] = new Array(numFrames).fill(0);
    const curveY: number[] = new Array(numFrames).fill(0);

    for (let i = 0; i <= n; i++) {
        const binomialCoeff = comb(n, i);
        for (let j = 0; j < numFrames; j++) {
            const basis = binomialCoeff * Math.pow(1 - t_vals[j], n - i) * Math.pow(t_vals[j], i);
            curveX[j] += basis * points[i][0];
            curveY[j] += basis * points[i][1];
        }
    }

    return [curveX, curveY];
}

function generateStretchedCurve(
    start: [number, number],
    end: [number, number],
    numFrames: number,
    curveFactor: number = 0.5
): { Xframes: number[], Yframes: number[] } {
    const [x0, y0] = start;
    const [x1, y1] = end;

    // x offset
    let offsetX = Math.sign(x1 - x0) * Math.abs(x1 - x0) * curveFactor;
    if (x0 === x1) {
        offsetX = -Math.abs(y1 - y0) * curveFactor; // y offset
    }

    // offset control point
    const controlX = (x0 + x1) / 2 + offsetX;
    const controlY = Math.max(y0, y1) + curveFactor * Math.abs(x1 - x0) + Math.abs(y1 - y0) / 2;

    // 3 control points
    const bezierPoints = [
        [x0, y0], // start point
        [controlX, controlY], // control point
        [x1, y1]  // end point
    ];
    const [Xframes, Yframes] = bezierCurve(bezierPoints, numFrames);

    return { Xframes, Yframes };
}

const testCases: Array<[[number, number], [number, number], number, number]> = [
    [[0, 0], [5, 5], 60, 0.5],
    [[5, 0], [0, 5], 60, 0.5],  
    [[2, 0], [2, 5], 60, 0.5],   
];

test
testCases.forEach(([start, end, frames, factor]) => {
    const { Xframes, Yframes } = generateStretchedCurve(start, end, frames, factor);
    console.log(`Start: ${start} → End: ${end}`);
    console.log("Xframes:", Xframes);
    console.log("Yframes:", Yframes);
});

File: Python/bezier_curve.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.special import comb

def bezier_curve(points, num_frames):
    """ Generate Bezier curve trajectory """
    n = len(points) - 1  # Number of control points
    t_vals = np.linspace(0, 1, num_frames)**1.5  # Accelerated time distribution
    curve = np.zeros((num_frames, 2))

    for i in range(n+1):
        binomial_coeff = comb(n, i)  # Compute binomial coefficient
        curve += binomial_coeff * ((1 - t_vals) ** (n - i))[:, None] * (t_vals ** i)[:, None] * points[i]

    return curve[:, 0], curve[:, 1]

def generate_stretched_curve(start, end, num_frames, curve_factor=0.5):
    """ Generate a curved trajectory ensuring acceleration """
    x0, y0 = start
    x1, y1 = end

    # Determine the X offset direction of the control point
    offset_x = np.sign(x1 - x0) * abs(x1 - x0) * curve_factor
    if x0 == x1:
        offset_x = -abs(y1 - y0) * curve_factor  # Offset to the left for vertical direction

    # Determine the Y offset of the control point to ensure upward curvature
    control_x = (x0 + x1) / 2 + offset_x
    control_y = max(y0, y1) + curve_factor * abs(x1 - x0) + abs(y1 - y0)/2

    # Three key points of the Bezier curve
    bezier_points = np.array([
        [x0, y0],  # Start point
        [control_x, control_y],  # Control point
        [x1, y1]   # End point
    ])

    # Generate Bezier curve
    x_vals, y_vals = bezier_curve(bezier_points, num_frames)

    return x_vals, y_vals

# Example: different test cases
test_cases = [
    ((0, 0), (5, 5), 60, 0.5),   # Start on the left, end on the right
    ((5, 0), (0, 5), 60, 0.5),   # Start on the right, end on the left
    ((2, 0), (2, 5), 60, 0.5),   # Vertical movement (should shift left)
]

# Plot all test trajectories
plt.figure(figsize=(6, 6))
for start, end, frames, factor in test_cases:
    x_vals, y_vals = generate_stretched_curve(start, end, frames, factor)
    plt.plot(x_vals, y_vals, marker='o', linestyle='-', markersize=3, label=f"Start {start} → End {end}")

# Visualization
plt.xlabel("X Position")
plt.ylabel("Y Position")
plt.title("Stretched Curve with Directional Offset")
plt.grid()
plt.legend()
plt.show()
